snuba/utils/streams/consumer.py

```diff
@@ -246,6 +287,12 @@ def revocation_callback(
                     on_revoke(partitions)
             finally:
                 for partition in partitions:
+                    # Staged offsets are deleted during partition revocation to
+                    # avoid committing offsets for partitions that are no
+                    # longer owned by this consumer.
+                    if partition in self.__staged_offsets:
+                        del self.__staged_offsets[partition]
+
```

Comment:
I cannot remember in which case we get a partition revoked and immediately reassigned to me. What happens if I get a partition revoked and immediately reassigned to me? Are we running into an issue if we drop the staged offsets there ?

Reply:
> I cannot remember in which case we get a partition revoked and immediately reassigned to me. What happens if I get a partition revoked and immediately reassigned to me?

From what I can tell (it's not documented), the partition revocation *always* happens immediately prior to reassignment, so yeah — this a pretty typical case.

> Are we running into an issue if we drop the staged offsets there?

The only case where this would be a problem is if a consumer did not commit as part of the revocation callback (the batching consumer always does.)

The other way to handle this and preserve staged offsets across rebalances (but only for owned partitions) would be to handle this in the assignment callback by taking the set difference of the two assignments, e.g. `revoked partitions = old assigned partition set - new assigned partition set` and dropping offsets for only those partitions.

Since we [force roll back offsets to the committed offset](https://github.com/getsentry/snuba/blob/9f67312d114ac871648228bc96c90f82dc8e0569/snuba/utils/streams/consumer.py#L211-L226) as part of the assignment process (this was largely to simplify the auto offset reset behavior as well as the synchronized consumer implementation, since offsets will always be available when an assignment exists), _not_ dropping the staged offsets could lead to a situation where:

1. consumer gets an assignment: {a:0, b: 0}
2. some messages are processed {a: [0-9], b: [0-19])
3. offsets {a: 10, b: 20} staged
4. another consumer joins and rebalance occurs _without commit_; consumer gets an assignment: {a:0} (rolled back to committed offset)
  a. this would drop {b: 20} from the staged offsets, leaving {a: 10}
5. staged offsets offsets ({a: 10}) committed
6. some messages are processed {a: [0-5]}
7. offsets {a: 5} staged
8. staged offsets are committed ({a: 5}) and _commit log offset is rolled back!_

There are no guarantees that the offset will move monotonically (and the synchronized consumer does its best to deal with this scenario as gracefully as possible), but this seems like a generally undesirable scenario.

Maybe a good balance would be to leave this as-is, but log a warning if unstaged offsets remain after the revocation callback (where commit _should_ be called under normal circumstances?)

Reply:
I like the log idea

Reply:
Implemented the log line here: 63e42d9f25a78e8eeb86b917d80ddfb6e4236b8e